### DISCUSSION TITLE
Why does C++ bag reading require headers?

### DISCUSSION THREAD

**Discussion by Cruzker** [foxglove/mcap#1255] [General]
Since MCAP files contain message definitions, you should be able to dynamically create structures to read the C++ messages without any header information involved using nested maps or something similar.

Is there a reason that currently this isn't done?

**Comment by jtbandes** [comment]
What do you mean by "without any header information involved"?

This sort of thing is definitely possible, see for example the dynamic protobuf reader here: https://mcap.dev/guides/cpp/protobuf#dynamically-read-fields
